### PR TITLE
Add tests for `GetInterfaceMap` on static interface methods

### LINKED ISSUE #90351: Exception on `MakeGenericMethod` called on a static interface method from `GetInterfaceMap`'s `TargetMethods`
### Description

**_This issue is for adding the tests, see Other information_**

When you use `MakeGenericMethod` on a static generic method, which comes from `GetInterfaceMap`'s `TargetMethods`, it throws an `InvalidOperationException`. It doesn't throw when you get the method directly from the type.

Gotta love static interface methods :)

Originally found by @Joe4evr on the C# discord, just thought I'd report it so it doesn't get lost.

### Reproduction Steps

```csharp
using System;
using System.Linq;
using System.Reflection;
public class C
{
    public static void Main()
    {
        var x1 = typeof(C1).GetMethod("I1.A", (BindingFlags)(-1));
        var x2 = typeof(C1).GetInterfaceMap(typeof(I1)).TargetMethods.Single((x) => x.Name == "I1.A"); //doesn't occur on InterfaceMethods
        Console.WriteLine(x1.MakeGenericMethod(typeof(int)).ContainsGenericParameters);
        Console.WriteLine(x2.MakeGenericMethod(typeof(int)).ContainsGenericParameters); //exception is here (on call to MakeGenericMethod)
    }
}

public interface I1
{
    static abstract void A<T>();
}

public class C1 : I1
{
    static void I1.A<T>() { } //doesn't need to be an explicit implementation
}
```

### Expected behavior

Prints:
```
False
False
```

### Actual behavior

Prints:
```
False
Exception
System.InvalidOperationException: Void I1.A[T]() is not a GenericMethodDefinition. MakeGenericMethod may only be called on a method for which MethodBase.IsGenericMethodDefinition is true.
   at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation)
   at C.Main()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

I tested this in sharplab.io
I have also reproduced locally on macOS 13.4.1 x64 with .NET 7.0.10

### Other information

The above described issue was fixed with other changes. This issue is now for adding tests to ensure we do not regress.
I also found another instance of the original issue: https://github.com/dotnet/runtime/issues/90351#issuecomment-1676179889.

### PR DESCRIPTION
**Pull request by hamarb123**
Fixes https://github.com/dotnet/runtime/issues/90351.
Includes the currently single failing case from the comment https://github.com/dotnet/runtime/issues/90351#issuecomment-1676179889 - therefore I propose that all (/ or at least most) of the combinations are useful since they actually found a bug.
This probably can't be merged until it's fixed, or maybe that just needs to be disabled for now, but that's alright lol.

The failing case is the case which checks `C2Implicit<string>`'s implementations of `I1<string>.G<T>`. Previously, until recently when some other unrelated (to this) changes to static interface methods were made, many more cases were broken.

Related failure issues, which have had their tests disabled:
- https://github.com/dotnet/runtime/issues/90863
- https://github.com/dotnet/runtime/issues/91027

cc @jkotas

### CODE REVIEW

**Comment by jkotas** [comment]
>maybe that just needs to be disabled for now, but that's alright lol.

I would recommend disabling the test against a new issue.

**Comment by hamarb123** [comment]
> I would recommend disabling the test against a new issue.

Is there a proper way to do this, or do I just need to skip the test manually? I ask because it's using an interesting setup, where it returns an `IEnumerable<object[]>` using `yield` as data for a `[Theory]`; presumably there's no way to do an `[ActiveIssue]` for just 1 of those cases?

**Comment by jkotas** [comment]
Comment out the failing case and add `// [ActiveIssue("https://github.com/dotnet/runtime/issues/12345")]` in front of it.

**Comment by hamarb123** [comment]
I also need to exclude some tests on mono by the looks of it, I will add an `[ActiveIssue]` for that when I make an issue for that too.

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.DUPLICATE_EXTENTS_DATA.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        // https://learn.microsoft.com/windows/win32/api/winioctl/ns-winioctl-duplicate_extents_data
+        internal struct DUPLICATE_EXTENTS_DATA
+        {
+            internal IntPtr FileHandle;
+            internal long SourceFileOffset;
+            internal long TargetFileOffset;
+            internal long ByteCount;
+        }
+    }
+}
```
Comment (hamarb123) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.DUPLICATE_EXTENTS_DATA.cs:19]: what, why is this here lol

File: src/libraries/System.Runtime/tests/System/Type/TypeTests.cs [hunk 2]
```diff
@@ -1070,6 +1097,148 @@ public static IEnumerable<object[]> GetInterfaceMap_TestData()
...
+                    {
+                        level2GenericInterface,
+                        classType.Type,
+                        new Tuple<MethodInfo, MethodInfo>[0]
+                    };
+                }
+
+                // Check we have the expected implementation for the level 3 interfaces (abstract explicit implementations - I1.M and I1.G methods)
+                bool isMono = typeof(object).Assembly.GetType("Mono.RuntimeStructs") != null; // [ActiveIssue("https://github.com/dotnet/runtime/issues/TODO")]
+                if (!isMono && classType.Index >= 3)
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/Type/TypeTests.cs:1216]:
```suggestion
[ActiveIssue("https://github.com/dotnet/runtime/issues/TODO")]
                if (!PlatformDetection.IsMonoRuntime && classType.Index >= 3)
```

**Review by trylek — APPROVED** [review]
Looks great to me, thank you!

**Comment by jkotas** [comment]
If there is nothing else to do, please flip the PR from Draft to Ready to review.

**Comment by hamarb123** [comment]
> If there is nothing else to do, please flip the PR from Draft to Ready to review.

I haven't opened the other issue yet, if you're alright with merging it without that issue then feel free, or if you wanted to open it that would be great since I'm struggling to run mono locally and therefore write a useful reproduction.

**Comment by jkotas** [comment]
For the tracking issue, it is fine to just link to the test that is disabled. You do not need to go the extra mile to extra the test into a small repro.

**Comment by jkotas** [comment]
Native aot compiler fails on the new tests with:
```
Process terminated. Assertion failed.
   at Internal.TypeSystem.MetadataVirtualMethodAlgorithm.ResolveInterfaceMethodToVirtualMethodOnType(MethodDesc interfaceMethod, MetadataType currentType) in /_/src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs:line 615
   at ILCompiler.LazyGenericsSupport.GraphBuilder.LookForVirtualOverrides(EcmaMethod method) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/GraphBuilder.cs:line 129
   at ILCompiler.LazyGenericsSupport.GraphBuilder..ctor(EcmaModule assembly) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/GraphBuilder.cs:line 62
   at ILCompiler.LazyGenericsSupport.CycleInfoHashtable.CreateValueFromKey(EcmaModule key) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 52
   at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
   at ILCompiler.LazyGenericsSupport.GenericCycleDetector.FormsCycle(TypeSystemEntity entity, ModuleCycleInfo& cycleInfo) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 194
   at ILCompiler.LazyGenericsSupport.GenericCycleDetector.DetectCycle(TypeSystemEntity owner, TypeSystemEntity referent) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 230
   at Internal.IL.ILImporter.GetMethodEntrypoint(MethodDesc method) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 257
   at Internal.IL.ILImporter.ImportCall(ILOpcode opcode, Int32 token) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 642
   at Internal.IL.ILImporter.ImportBasicBlock(BasicBlock basicBlock) in /_/src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs:line 424
   at Internal.IL.ILImporter.ImportBasicBlocks() in /_/src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs:line 308
   at Internal.IL.ILImporter.Import() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 166
   at ILCompiler.ILScanner.CompileSingleMethod(ScannedMethodNode methodCodeNodeNeedingCode) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 122
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`2.<ForWorker>b__1(RangeWorker& currentWorker, Int64 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
```
I am going to push a fix for this crash.

**Review by jkotas — APPROVED** [review]
Thank you!
